pandas/tools/tests/test_merge.py

```diff
@@ -2522,6 +2523,120 @@ def test_concat_multiindex_with_tz(self):
         result = concat([df, df])
         tm.assert_frame_equal(result, expected)
 
+    def test_concat_NaT_dataframes_all_NaT_axis_0(self):
+        # GH 12396
+        expect = pd.DataFrame([pd.NaT, pd.NaT, pd.NaT], index=[0, 1, 0])
+
+        # non-timezone aware
+        first = pd.DataFrame([[pd.NaT], [pd.NaT]])
+        second = pd.DataFrame([[pd.NaT]])
+
+        result = pd.concat([first, second], axis=0)
+        assert_frame_equal(result, expect)
+
+        # one side timezone-aware
+        dtype = DatetimeTZDtype('ns', tz='UTC')
+        first = pd.DataFrame([[pd.NaT], [pd.NaT]], dtype=dtype)
+
+        result = pd.concat([first, second], axis=0)
+        # upcasts for mixed case
+        assert_frame_equal(result, expect, check_dtype=False)
```

Comment:
make a new expect here that matches exactly what you want rather, so ``assert_frame_equal(result, expected)`` works & checks dtypes